staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func (rdm *resourceDiscoveryManager) AddGroupVersion(groupName string, value metav1.APIVersionDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.addGroupVersionLocked(groupName, value)
+}
+
+func (rdm *resourceDiscoveryManager) addGroupVersionLocked(groupName string, value metav1.APIVersionDiscovery) {
+	klog.Infof("Adding GroupVersion %s %s to ResourceManager", groupName, value.Version)
+
+	if rdm.apiGroups == nil {
+		rdm.apiGroups = make(map[string]*metav1.APIGroupDiscovery)
+	}
+
+	if existing, groupExists := rdm.apiGroups[groupName]; groupExists {
+		// If this version already exists, replace it
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		rdm.apiGroups[groupName] = &metav1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []metav1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroupNames = append(rdm.apiGroupNames, groupName)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		delete(rdm.apiGroups, group.Name)
+		for i := range rdm.apiGroupNames {
+			if rdm.apiGroupNames[i] == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+	for i := range rdm.apiGroupNames {
+		if rdm.apiGroupNames[i] == groupName {
+			rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			break
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) WebService() *restful.WebService {
+	mediaTypes, _ := negotiation.MediaTypesForSerializer(rdm.serializer)
+	ws := new(restful.WebService)
+	ws.Path(DiscoveryEndpointRoot)
+	ws.Doc("get available API groupversions and resources")
+
+	ws.Route(ws.GET("/v2").To(func(req *restful.Request, resp *restful.Response) {
```

Comment:
v2 stands out here.  the API is metav1, right?

Reply:
right. changed to v1